File: apps/web/src/server/jobs/booking-cancelled.ts
/**
 * `booking.cancelled` — tell the *other* party that a booking was cancelled.
 * The actor already saw the result on screen, so only the counterparty is
 * notified; the job records who cancelled and the recipient is derived from it.
 *
 * Reached as a QStash delivery to `POST /api/jobs/booking.cancelled`; the
 * dispatch layer (`run.ts`) has already validated the payload.
 */

import {
  type BookingCancelledJob,
  cancelNotificationRecipient,
  isDemoOrganizerId,
} from '@repo/contracts'

import { issueLoginLink } from '../auth/login-link'
import { getPostHog } from '../posthog'
import type { JobsEnv } from './env'
import { cabinetSlotPath, loginLinkUrl, organizerPageUrl } from './links'
import { getNotificationContext } from './notification-context'
import { sendMessage } from './telegram/client'
import {
  bookingCancelledForGuest,
  bookingCancelledForOrganizer,
  notificationLocale,
} from './telegram/templates'

import 'server-only'

export async function handleBookingCancelled(
  env: JobsEnv,
  data: BookingCancelledJob,
): Promise<void> {
  const { bookingId, cancelledBy } = data

  const context = await getNotificationContext(bookingId)
  if (!context) {
    console.warn(`[booking.cancelled] booking ${bookingId} no longer exists — skipping`)
    return
  }

  if (isDemoOrganizerId(context.organizer.id)) {
    console.warn(`[booking.cancelled] refusing to notify the demo organizer (${bookingId})`)
    return
  }

  const recipient = cancelNotificationRecipient(cancelledBy)
  const ph = getPostHog()

  if (recipient === 'organizer') {
    const token = await issueLoginLink(context.organizer.id, cabinetSlotPath(context.slot.id))

    const message = bookingCancelledForOrganizer(
      context,
      loginLinkUrl(env.appUrl, token),
      notificationLocale(recipient, context),
    )

    await sendMessage(env.telegramBotToken, {
      chatId: context.organizer.messengerId,
      text: message.text,
      button: message.button,
    })
    ph?.capture({
      distinctId: context.organizer.id,
      event: 'notification_sent',
      properties: { queue: 'booking.cancelled', recipient, bookingId },
    })
    return
  }

  const message = bookingCancelledForGuest(
    context,
    organizerPageUrl(env.appUrl, context.organizer.slug),
    notificationLocale(recipient, context),
  )

  await sendMessage(env.telegramBotToken, {
    chatId: context.booking.guestMessengerId,
    text: message.text,
    button: message.button,
  })
  ph?.capture({
    distinctId: context.organizer.id,
    event: 'notification_sent',
    properties: { queue: 'booking.cancelled', recipient, bookingId },
  })
}

File: apps/api-go/internal/jobs/booking_cancelled.go
package jobs

import (
	"context"

	"api-go/internal/auth"
	"api-go/internal/contracts"
	"api-go/internal/db"
	"api-go/internal/logx"
)

// booking.cancelled — tell the *other* party that a booking was
// cancelled: the actor already saw the result on screen, so only the
// counterparty is notified; the job records who cancelled and the
// recipient is derived from it.

func HandleBookingCancelled(ctx context.Context, env Env, job contracts.BookingCancelledJob) error {
	booking, slot, service, organizer, err := db.GetBookingChain(ctx, job.BookingID)
	if err != nil {
		return err
	}
	if booking == nil {
		logx.Info("booking no longer exists — skipping", map[string]any{
			"queue": contracts.QueueBookingCancelled, "bookingId": job.BookingID,
		})
		return nil
	}
	view := BookingView{Booking: *booking, Slot: *slot, Service: *service, Organizer: *organizer}

	if contracts.IsDemoOrganizerID(organizer.ID) {
		logx.Info("refusing to notify the demo organizer", map[string]any{
			"queue": contracts.QueueBookingCancelled, "bookingId": job.BookingID,
		})
		return nil
	}

	recipient := contracts.CancelNotificationRecipient(job.CancelledBy)

	if recipient == contracts.RecipientOrganizer {
		token, err := auth.IssueLoginLink(ctx, organizer.ID, CabinetSlotPath(slot.ID))
		if err != nil {
			return err
		}
		locale := NotificationLocale(contracts.RecipientOrganizer, view)
		message := BookingCancelledForOrganizer(view, LoginLinkURL(env.AppURL, token), locale)
		return SendMessage(ctx, env.TelegramBotToken, organizer.MessengerID, message.Text, message.Button)
	}

	locale := NotificationLocale(contracts.RecipientGuest, view)
	message := BookingCancelledForGuest(view, OrganizerPageURL(env.AppURL, organizer.Slug), locale)
	return SendMessage(ctx, env.TelegramBotToken, booking.GuestMessengerID, message.Text, message.Button)
}
